### DISCUSSION TITLE
Help with ejabberd (eCS) Resource Utiliztion

### DISCUSSION THREAD

**Discussion by WalManas** [processone/ejabberd#3925] [Questions] [answered]
Hello,

I am a graduate student conducting a comprehensive survey of open-source messaging middleware as a research for my thesis. I have a few questions that I could not find answers for in the documentation.

1. Can the amount of disk available for mnesia storage be capped to some amount, or maybe percentage?
2. Does ejabberd impose any quota on CPU threads utilization? i.e. does it force some sort of fair use by limiting the amount of time/threads a user can use?
3. For mod_pubsub, what happens when the max_items_node limit is exhausted? Does ejabberd drop older messages to make space for new ones?

Thanks in advance for any help,

Wael.

**Comment by badlop** [comment, marked as answer]
> 1. Can the amount of disk available for mnesia storage be capped to some amount, or maybe percentage?


ejabberd itself does not include such a configuration. I took a look at Mnesia documentation, but found nothing like that in https://www.erlang.org/doc/apps/mnesia/users_guide.html

I remember that the ejabberd log files sometimes reported an Erlang warning that the disk was getting almost full. This was back when my disk was almost full of other stuff, and erlang detected that risky situation... But I couldn't reproduce this now, no documentation, and in any case it was just warning reports.

> 2. Does ejabberd impose any quota on CPU threads utilization? i.e. does it force some sort of fair use by limiting the amount of time/threads a user can use?

For generic erlang process counter:
https://github.com/processone/ejabberd/blob/master/ejabberdctl.cfg.example#L82

For user/account limits... Each xmpp account is represented in ejabberd as an erlang process

There are some ways to restrict xmpp accounts usage (and abuse):
https://docs.ejabberd.im/admin/configuration/basic/#shapers
https://docs.ejabberd.im/admin/configuration/basic/#limiting-opened-sessions


> 3. For mod_pubsub, what happens when the max_items_node limit is exhausted? Does ejabberd drop older messages to make space for new ones?

That option was added in https://github.com/processone/ejabberd/commit/29dcc9b94ccfd514cf388979e7210d01cb97d5f4 looking at that commit, it is used only when running the API command delete_expired_pubsub_items

**Reply by WalManas** [reply]
Hi @badlop,

Thanks for your response. I did not understand how max_items_node and delete_expired_pubsub_items work!

After max_items_node is met, does the node accepts newly produced messages at the expense of marking the oldest ones as expired, which then are deleted once delete_expired_pubsub_items is called? If yes, are the expired messages accessible to consumers or they are moved to some dead-letter queue?

Thanks again!

**Reply by badlop** [reply]
It works exactly as I explained, and the key word is "only":

> That option was added in https://github.com/processone/ejabberd/commit/29dcc9b94ccfd514cf388979e7210d01cb97d5f4 looking at that commit, it is used **only** when running the API command delete_expired_pubsub_items

Nothing special is done at all while nodes are being published, requested, etc. When the admin runs `delete_expired_pubsub_items`, and **only** in that case, the option `max_items_node` is read from the configuration table, and the nodes are traversed, checking which nodes are expired, and those are deleted.

**Reply by WalManas** [reply]
Thanks @badlop. I posted a couple other questions on the General discussion. I would appreciate your help with them if you can provide any.

Thanks again!
